yt/data_objects/static_output.py

```diff
@@ -721,42 +737,83 @@ def setup_deprecated_fields(self):
             added.append(("gas", old_name))
         self.field_info.find_dependencies(added)
 
-    def _setup_coordinate_handler(self):
-        kwargs = {}
-        if isinstance(self.geometry, tuple):
-            self.geometry, ordering = self.geometry
-            kwargs["ordering"] = ordering
+    def _setup_coordinate_handler(self, axis_order: Optional[AxisOrder]) -> None:
+        # backward compatibility layer:
+        # turning off type-checker on a per-line basis
+
+        if isinstance(self.geometry, tuple):  # type: ignore [unreachable]
+            issue_deprecation_warning(  # type: ignore [unreachable]
+                f"Dataset object {self} has a tuple for its geometry attribute. "
+                "This is interpreted as meaning the first element is the actual geometry string, "
+                "and the second represents an arbitrary axis order. "
+                "This will stop working in a future version of yt.\n"
+                "If you're loading data using yt.load_* functions, "
+                "you should be able to clear this warning by using the axis_order keyword argument.\n"
+                "Otherwise, if your code relies on this behaviour, please reach out and open an issue:\n"
+                "https://github.com/yt-project/yt/issues/new",
+                since="4.2",
+                stacklevel=2,
+            )
+            self.geometry, axis_order = self.geometry
+        if type(self.geometry) is str:
+            issue_deprecation_warning(
+                f"Dataset object {self} has a raw string for its geometry attribute. "
+                "In yt>=4.2, a yt.geometry.geometry_enum.Geometry member is expected instead. "
+                "This will stop working in a future version of yt.\n",
+                since="4.2",
+                stacklevel=2,
+            )
+            self.geometry = Geometry(self.geometry.lower())
         if isinstance(self.geometry, CoordinateHandler):
-            # I kind of dislike this.  The geometry field should always be a
-            # string, but the way we're set up with subclassing, we can't
-            # mandate that quite the way I'd like.
-            self.coordinates = self.geometry
-            return
-        elif callable(self.geometry):
```

Comment:
Me neither. I _think_ it was written in support for some frontends that may have been refactored since then. On the off chance that anyone out there is relying on this, I could indeed add a deprecation period.

Reply:
I'm not big on supporting code that _might_ be useful. How about deprecate this with a message along the lines of 

>if your code relies on this behaviour, please reach out and open an issue:\nhttps://github.com/yt-project/yt/issues/new

So we don't immediately break and have a chance to go back in case of problem ?

Reply:
Additional thoughts on this:
I have serious doubts this ever worked properly, because it leaves `self.geometry` as a class, so even if `self.coordinates` is setup correctly,  `self.geometry` might be compared to strings at any later point, which is never acknowledged (because all downstream code assumes it's a string already). This means that the result of an if/elif/else   will depend on how it is implemented. For illustration, let's assume there are exactly three geometries available (say `"cartesian"`, `"polar"`, and `"geographic"`), one might write a 3 way if tree as
```python
# case 1
if self.geometry == "cartesian":
    branch_cartesian()
if self.geometry == "polar":
    branch_polar()
else:  # "geographic" is implicitly assumed
    # this branch is selected !
    branch_geographic()
```
*or* as
```python
# case 2
if self.geometry == "cartesian":
    branch_cartesian()
if self.geometry == "geographic":
    branch_geographic()
else:  # "polar" is implicitly assumed
    # this branch is selected !
    branch_polar()
```

so, if `self.geometry` is allowed to break the rule ("must be a str"), every piece of code that implements geometry-flexible behaviour must take it into account, but it is not the case.

Additionally, this special casing is not _necessary_ to support dynamically constructed geometry handler classes.
One way to implement it with type safety would be to
- add a member to the `Geometry` enum (e.g. `Geometry.DYNAMIC`)
- override the `_setup_geometry_handler` method in the frontend

This PR makes it so that *if* and *when* we want to do that, we will _have_ to revise all parts of the case that implement geometry-flexibility, so that the kind of bugs I just described is detected at type-check time !

So I'm still thinking that deprecating it with a message like "need this ? get in touch !" is the way to go.
Thoughts ?

Reply:
> So I'm still thinking that deprecating it with a message like "need this ? get in touch !" is the way to go.
Thoughts ? 

That makes sense to me. @matthewturk ?

Reply:
Sure!

Reply:
thank you guys, this should be resolved with f4edf9d3011dc1c01c5d21cd825086e59700c4ff